core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala

```diff
@@ -114,175 +170,183 @@ trait RawlsBillingProjectComponent {
+    def withBillingAccount(billingAccount: Option[RawlsBillingAccountName]): RawlsBillingProjectQuery =
+      query.filter(_.billingAccount === billingAccount.map(_.value))
 
-      BillingProjectSpendExport(RawlsBillingProjectName(projectRecord.projectName), billingAccount, table)
-    }
+    def withCreationStatus(status: CreationStatus): RawlsBillingProjectExtensions =
+      query.filter(_.creationStatus === status.toString)
+
+    def withCreationStatuses(statuses: Iterable[CreationStatus]): RawlsBillingProjectQuery =
+      query.filter(_.creationStatus.inSetBind(statuses.map(_.toString)))
+
+    def withServicePerimeter(perimeter: Option[ServicePerimeterName]): RawlsBillingProjectQuery =
+      query.filter(_.servicePerimeter === perimeter.map(_.value))
+
+
+    // setters
+    def setServicePerimeter(servicePerimeter: Option[ServicePerimeterName]): WriteAction[Int] =
+      query
+        .map(_.servicePerimeter)
+        .update(servicePerimeter.map(_.value))
+
+    def setGoogleProjectNumber(googleProjectNumber: Option[GoogleProjectNumber]): WriteAction[Int] =
+      query
+        .map(_.googleProjectNumber)
+        .update(googleProjectNumber.map(_.value))
+
+    def setCreationStatus(status: CreationStatus, message: Option[String] = None): WriteAction[Int] =
+      query
+        .map(p => (p.creationStatus, p.message))
+        .update((status.toString, message))
+
+    def setMessage(message: Option[String]): WriteAction[Int] =
+      query.map(_.message).update(message)
+
+    def setBillingProjectSpendConfiguration(datasetName: Option[BigQueryDatasetName],
+                                            tableName: Option[BigQueryTableName],
+                                            datasetGoogleProject: Option[GoogleProject]
+                                           ): WriteAction[Int] =
+      query
+        .map(bp => (bp.spendReportDataset, bp.spendReportTable, bp.spendReportDatasetGoogleProject))
+        .update(datasetName.map(_.value), tableName.map(_.value), datasetGoogleProject.map(_.value))
+
+    def setInvalidBillingAccount(isInvalid: Boolean): WriteAction[Int] =
+      query.map(_.invalidBillingAccount).update(isInvalid)
+
+    def setBillingAccount(billingAccount: Option[RawlsBillingAccountName],
+                          userSubjectId: RawlsUserSubjectId): ReadWriteAction[Int] =
+      for {
+        billingProjects <- query.read
+        count <- query.map(_.billingAccount).update(billingAccount.map(_.value))
+        // Record each billing account change
+        // - so the `WorkspaceBillingAccountActor` can synchronise the changes with google
+        // - to keep an audit log of billing account changes
+        _ <- DBIO.sequence(billingProjects.map { project =>
+          BillingAccountChanges.create(
+            project.projectName,
```

Comment:
Yeah, this is the compromise we made by not doing this as a database hook. As you say, since this is in a transaction it should be fine. Then again, there's nothing stopping someone updating the billing project table directly.